integration/simulation/transaction_injector.go

```diff
@@ -284,37 +310,49 @@ func (m *TransactionInjector) rndL2NodeClient() *obscuroclient.Client {
 	return m.l2Clients[rand.Intn(len(m.l2Clients))] //nolint:gosec
 }
 
-// NewL2Transfer creates an enclave.L2Tx of type enclave.TransferTx
-func NewL2Transfer(from common.Address, dest common.Address, amount uint64) types.TxData {
-	txData := core.L2TxData{Type: core.TransferTx, From: from, To: dest, Amount: amount}
-	return NewL2Tx(txData)
+func NewObscuroTransferTx(from wallet.Wallet, dest common.Address, amount uint64, client *obscuroclient.Client) types.TxData {
+	data := erc20contractlib.CreateTransferTxData(dest, amount)
+	return newTx(data, NextNonce(client, from))
+}
+
+func NewObscuroWithdrawalTx(from wallet.Wallet, amount uint64, client *obscuroclient.Client) types.TxData {
+	transferERC20data := erc20contractlib.CreateTransferTxData(evm.WithdrawalAddress, amount)
+	return newTx(transferERC20data, NextNonce(client, from))
 }
 
-// NewL2Withdrawal creates an enclave.L2Tx of type enclave.WithdrawalTx
-func NewL2Withdrawal(from common.Address, amount uint64) types.TxData {
-	txData := core.L2TxData{Type: core.WithdrawalTx, From: from, Amount: amount}
-	return NewL2Tx(txData)
+func NewCustomObscuroWithdrawalTx(amount uint64) types.TxData {
+	transferERC20data := erc20contractlib.CreateTransferTxData(evm.WithdrawalAddress, amount)
+	return newTx(transferERC20data, 1)
 }
 
-// NewL2Tx creates an enclave.L2Tx.
-//
-// A random nonce is used to avoid hash collisions. The enclave.L2TxData is encoded and stored in the transaction's
-// data field.
-func NewL2Tx(data core.L2TxData) types.TxData {
-	// We should probably use a deterministic nonce instead, as in the L1.
-	nonce := rand.Intn(math.MaxInt) //nolint:gosec
+func newTx(data []byte, nonce uint64) types.TxData {
+	return &types.LegacyTx{
+		Nonce:    nonce,
+		Value:    common.Big0,
+		Gas:      1_000_000,
+		GasPrice: common.Big0,
+		Data:     data,
+		To:       &evm.Erc20ContractAddress,
+	}
+}
 
-	enc, err := rlp.EncodeToBytes(data)
+func readNonce(cl *obscuroclient.Client, a common.Address) uint64 {
+	var result uint64
+	err := (*cl).Call(&result, obscuroclient.RPCNonce, a)
 	if err != nil {
-		// TODO - Surface this error properly.
 		panic(err)
 	}
+	return result
+}
 
-	return &types.LegacyTx{
-		Nonce:    uint64(nonce),
-		Value:    big.NewInt(1),
-		Gas:      1,
-		GasPrice: big.NewInt(1),
-		Data:     enc,
+func NextNonce(cl *obscuroclient.Client, w wallet.Wallet) uint64 {
+	// only returns the nonce when the previous transaction was recorded
+	for {
```

Comment:
This strat might backfire when issuing txs to different nodes. ( one node might have processed the tx and increased the nonce but another node mightn't )

Typically the nonce management is always at the client (issuer side)

Reply:
All nodes reach consensus eventually.

This is a hack, which works for now. Once we have the receipts endpoint, this logic can be replaced with something that checks that the previous transaction was processed correctly.